tools/linker/CoreTypeMapStep.cs

```diff
@@ -20,27 +20,101 @@
 namespace MonoTouch.Tuner {
 
 	// This class is shared between Xamarin.Mac and Xamarin.iOS
-	public class CoreTypeMapStep : TypeMapStep {
-		HashSet<TypeDefinition> cached_isnsobject = new HashSet<TypeDefinition> ();
-		Dictionary<TypeDefinition, bool?> isdirectbinding_value = new Dictionary<TypeDefinition, bool?> ();
+	public class CoreTypeMapStep :
+#if NET
+		ConfigurationAwareStep
+#else
+		TypeMapStep
+#endif
+	{
+
+#if NET
+		protected override string Name { get; } = "CoreTypeMap";
+		protected override int ErrorCode { get; } = 2381;
+
+		Profile Profile => new Profile (Configuration);
+
+		Dictionary<AssemblyDefinition, bool> _transitivelyReferencesProduct = new Dictionary<AssemblyDefinition, bool> ();
+		bool TransitivelyReferencesProduct (AssemblyDefinition assembly)
+		{
+			if (_transitivelyReferencesProduct.TryGetValue (assembly, out bool result))
+				return result;
+
+			if (Profile.IsProductAssembly (assembly)) {
+				_transitivelyReferencesProduct.Add (assembly, true);
+				return true;
+			}
+
+			foreach (var reference in assembly.MainModule.AssemblyReferences) {
+				if (!Configuration.AssembliesByName.TryGetValue (reference.Name, out AssemblyDefinition resolvedReference))
```

Comment:
That works because you enforced the (assembly) _weak_ name to be unique earlier.
Not sure it will work, unless everything was resolved (to latest version) before the linker is given the list (but then that check would not be needed)

Reply:
The linker also expects weak assembly names to be unique. You're right that the check was unnecessary - I forgot that GetLoadedAssembly is available.

This still relies on the workaround to load assemblies up-front: https://github.com/xamarin/xamarin-macios/blob/882dde132a0b893c1e929969d7472e93ec7745c8/tools/dotnet-linker/Steps/CollectAssembliesStep.cs#L19

I thought there was a way to get rid of this by resolving assemblies ourselves, but that doesn't appear to be the case - I filed https://github.com/mono/linker/issues/2073 about it.